README.md
# Single Step KDF (NIST SP 800-56C)

This is an implementation of the single-step key derivation function as described
in [NIST SP 800-56C revision 1, chapter 4](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr1.pdf).
It is an unopinionated approach towards the subject, allowing all 3 options (message digest, hmac and kmac) as H
function and leaving open the exact format of the `fixedInfo` parameter.

[![Maven Central](https://img.shields.io/maven-central/v/at.favre.lib/singlestep-kdf)](https://mvnrepository.com/artifact/at.favre.lib/singlestep-kdf)
[![Github Actions](https://github.com/patrickfav/singlestep-kdf/actions/workflows/build_deploy.yml/badge.svg)](https://github.com/patrickfav/singlestep-kdf/actions)
[![Javadocs](https://www.javadoc.io/badge/at.favre.lib/singlestep-kdf.svg)](https://www.javadoc.io/doc/at.favre.lib/singlestep-kdf)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=patrickfav_singlestep-kdf&metric=coverage)](https://sonarcloud.io/summary/new_code?id=patrickfav_singlestep-kdf)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=patrickfav_singlestep-kdf&metric=security_rating)](https://sonarcloud.io/summary/new_code?id=patrickfav_singlestep-kdf)
[![Technical Debt](https://sonarcloud.io/api/project_badges/measure?project=patrickfav_singlestep-kdf&metric=sqale_index)](https://sonarcloud.io/summary/new_code?id=patrickfav_singlestep-kdf)

This is a standalone, lightweight, simple to use, fully tested and stable implementation in Java. The code is compiled
with [Java 7](https://en.wikipedia.org/wiki/Java_version_history#Java_SE_7) to be compatible with most [
_Android_](https://www.android.com/) versions as well as normal Java applications.

## Quickstart

Add dependency to your `pom.xml`:

```xml
<dependency>
    <groupId>at.favre.lib</groupId>
    <artifactId>singlestep-kdf</artifactId>
    <version>{latest-version}</version>
</dependency>
```

A very simple example:

```java
// a shared secret provided by your protocol
byte[] sharedSecret = ...
// a salt; if you don't have access to a salt, use salt-less SingleStepKdf.fromSha256() or similar
byte[] salt = ...
// other info to bind the key to the context, see the NIST spec for more detail
byte[] otherInfo = "macKey".getBytes();
byte[] keyMaterial = SingleStepKdf.fromHmacSha256().derive(sharedSecret, 32, salt, otherInfo);
SecretKey secretKey = new SecretKeySpec(keyMaterial, "AES");
```

## Full Example

NIST 800-56C specifies the KDF with an H-function which may be a [hash](https://en.wikipedia.org/wiki/Cryptographic_hash_function), [HMAC](https://en.wikipedia.org/wiki/HMAC) or [KMAC](https://keccak.team/index.html).

### Using with Message Digest (Option 1)

Using Option 1, i.e. `H(x) = hash(x)`, where hash is an approved hash function. Note that when you use this option, the
salt parameter is not supported. If you want to incorporate a salt just include it into the `fixedInfo` parameter. **If
you have no specific reason for choosing Option 1, I would always prefer Option 2 (HMAC) over this one.**

```java
// a shared secret provided by your protocol
byte[] sharedSecret = Bytes.random(16).array();
// fixed info to bind the key to the context
byte[] fixedInfo = "macKey".getBytes();
// salt is not directly supported with message digest, you may include it in fixedInfo
byte[] keyMaterial = SingleStepKdf.fromSha256().derive(sharedSecret, 32, fixedInfo);
SecretKey secretKey = new SecretKeySpec(keyMaterial, "AES");
```

### Using with HMAC (Option 2)

Using Option 2, i.e. `H(x) = HMAC-hash(salt, x)`, where HMAC-hash is an implementation of the HMAC algorithm (as defined
in FIPS 198) employing an approved hash function. A salt which serves as the HMAC key, and x (the input to H) is a bit
string that serves as the HMAC "message". This library can use any `Mac` implementation. If no salt is provided, an
empty array is internally initialized.

```java
byte[] keyMaterial = SingleStepKdf.fromHmacSha256().derive(sharedSecret, 32, salt, fixedInfo);
        SecretKey secretKey = new SecretKeySpec(keyMaterial, "AES");
```

### Using with KMAC (Option 3)

KMAC is a MAC using [SHA-3/Keccak](https://en.wikipedia.org/wiki/SHA-3). Unlike SHA-1 and
SHA-2, [Keccak](http://keccak.noekeon.org/) does not have
the [length-extension weakness](https://en.wikipedia.org/wiki/Length_extension_attack), hence does not need the HMAC
nested construction. Instead, MAC computation can be performed by simply prepending the message with the key. Java has
an SHA-3 implementation [since version 9](https://openjdk.java.net/jeps/287). This library is designed to support Java
7, so no default implementation is present for KMAC. It is probably quite easy to implement it yourself using either
the `HFunctionFactory.MacFactory` or implementing yourself with `HFunction`.

### Using custom Message Digest / HMAC implementation

Default implementation exists for the h-function which can be used for any `MessageDigest` or `Mac`. It is also possible to implement it from scratch by using the `HFunction` interface. A factory is used to generate instances. This can be used like this:

```java
// create instance with sha1 as backing hash function
SingleStepKdf sha1Kdf = SingleStepKdf.from(new HFunctionFactory.Default.DigestFactory("SHA-1"));
// create instance with HMAC-SHA1 as backing h function
SingleStepKdf hmacSha1Kdf = SingleStepKdf.from(new HFunctionFactory.Default.DigestFactory("HmacSHA1"));
```

### How to use the Fixed-Info Parameter

A bit string of context-specific data that is appropriate for the relying on key-establishment scheme. As its name
suggests, the value of `FixedInfo` does not change during the execution of the process.

`FixedInfo` may, for example, include appropriately formatted representations of the values of salt and/or the output
length. The inclusion of additional copies of the values of salt and the output length in `FixedInfo` would ensure that
each block of derived keying material is affected by all the information conveyed in `OtherInput`.
See [SP 800-56A](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar2.pdf)
and [SP 800-56B](https://nvlpubs.nist.gov/nistpubs/specialpublications/nist.sp.800-56br1.pdf) for more detailed
recommendations concerning the format and content of `FixedInfo` (also known as OtherInfo in earlier versions of those
documents).

## Download

The artifacts are deployed to [jcenter](https://bintray.com/bintray/jcenter) and [Maven Central](https://search.maven.org/).

### Maven

Add dependency to your `pom.xml`:

```xml
<dependency>
    <groupId>at.favre.lib</groupId>
    <artifactId>singlestep-kdf</artifactId>
    <version>{latest-version}</version>
</dependency>
```

### Gradle

Add to your `build.gradle` module dependencies:

    compile group: 'at.favre.lib', name: 'singlestep-kdf', version: '{latest-version}'

### Local Jar

[Grab jar from the latest release.](https://github.com/patrickfav/singlestep-kdf/releases/latest)


## Description

The following is summarized and shortened version of [NIST 800-56C Rev1](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr1.pdf) chapter about the single-step KDF:

Single Step KDF specifies a family of approved key-derivation functions (KDFs) that are executed in a single step; The input to each specified KDF includes the shared secret generated during the execution of a key-establishment scheme specified in [SP 800-56A](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar2.pdf) or [SP 800-56B](https://nvlpubs.nist.gov/nistpubs/specialpublications/nist.sp.800-56br1.pdf), an indication of the desired bit length of the keying material to be output, and, perhaps, other information (as determined by the particular implementation of the key-establishment scheme and/or key-derivation function).

Implementations of these one-step KDFs depend upon the choice of an auxiliary function H, which can be either:
 
 1. an approved hash function, denoted as hash, as defined in [FIPS 180](https://nvlpubs.nist.gov/nistpubs/fips/nist.fips.180-4.pdf) or [FIPS 202](https://nvlpubs.nist.gov/nistpubs/fips/nist.fips.202.pdf); 
 2. HMAC with an approved hash function, hash, denoted as HMAC-hash, and defined in [FIPS 198](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.198-1.pdf); or 
 3. a KMAC variant, as defined in [SP 800-185](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf). H shall be chosen in accordance with the selection requirements specified in [800-56C Rev1/Section 7](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr1.pdf).

When an approved MAC algorithm (HMAC or KMAC) is used to define the auxiliary function H, it is permitted to use a known salt value as the MAC key. In such cases, it is assumed that the MAC algorithm will satisfy the following property (for each of its supported security strengths):

Given knowledge of the key `k`, and (perhaps) partial knowledge of a message `x` that includes an unknown substring `z`, it must be computationally infeasible to predict the (as-yet-unseen) value of `MAC(k, x, …)` with a probability of success that is a significant improvement over simply guessing either the MAC value or the value of `z`, even if one has already seen the values of `MAC(kj, xj, …)` for a feasible number of other `(kj, xj)` pairs, where each key `kj` is known and each (partially known) message `xj` includes the same unknown substring `z`, provided that none of the `(kj, xj)` pairs is identical to `(k, x)`.

This property is consistent with the use of the MAC algorithm as the specification of a family of pseudorandom functions defined on the appropriate message space and indexed by the choice of MAC key. Under Option 2 and Option 3 of the KDF specification below, the auxiliary function H is a particular selection from such a family.

### Two-Step Key Derivation Function

NIST 800-56C Rev1 also describes a two-step kdf with an extract and expand phase. The most prominent implementation of
it is [HKDF (RFC5869)](https://tools.ietf.org/html/rfc5869). A java implementation of it can
be [found here](https://github.com/patrickfav/hkdf).

### Test Vectors

Unfortunately it seems that the NIST did not provide any official test
vectors ([see this post](https://crypto.stackexchange.com/questions/64140/where-can-i-find-official-test-vectors-for-nist-sp-800-56c-r1-single-step-kdf)).
This implementation ist tested against
the [code snippets posted here](https://stackoverflow.com/questions/10879658/existing-implementations-for-nist-sp-800-56a-concatenation-key-derivation-functi/10971402#10971402).
Additionally, I released
my [own test vectors in the wiki](https://github.com/patrickfav/singlestep-kdf/wiki/NIST-SP-800-56C-Rev1:-Non-Official-Test-Vectors),
so you could test against another possible already existing implementation.

## Security Relevant Information

### OWASP Dependency Check

This project uses the [OWASP Dependency-Check](https://www.owasp.org/index.php/OWASP_Dependency_Check) which is a utility that identifies project dependencies and checks if there are any known, publicly disclosed, vulnerabilities against a [NIST database](https://nvd.nist.gov/vuln/data-feeds).
The build will fail if any issue is found.

### Digital Signatures

#### Signed Jar

The provided JARs in the GitHub release page are signed with my private key:

    CN=Patrick Favre-Bulle, OU=Private, O=PF Github Open Source, L=Vienna, ST=Vienna, C=AT
    Validity: Thu Sep 07 16:40:57 SGT 2017 to: Fri Feb 10 16:40:57 SGT 2034
    SHA1: 06:DE:F2:C5:F7:BC:0C:11:ED:35:E2:0F:B1:9F:78:99:0F:BE:43:C4
    SHA256: 2B:65:33:B0:1C:0D:2A:69:4E:2D:53:8F:29:D5:6C:D6:87:AF:06:42:1F:1A:EE:B3:3C:E0:6D:0B:65:A1:AA:88

Use the jarsigner tool (found in your `$JAVA_HOME/bin` folder) folder to verify.

#### Signed Commits

All tags and commits by me are signed with git with my private key:

    GPG key ID: 4FDF85343912A3AB
    Fingerprint: 2FB392FB05158589B767960C4FDF85343912A3AB

## Build

### Jar Sign

If you want to jar sign you need to provide a file `keystore.jks` in the
root folder with the correct credentials set in environment variables (
`OPENSOURCE_PROJECTS_KS_PW` and `OPENSOURCE_PROJECTS_KEY_PW`); alias is
set as `pfopensource`.

If you want to skip jar signing just change the skip configuration in the
`pom.xml` jar sign plugin to true:

    <skip>true</skip>

### Build with Maven

Use maven wrapper to create a jar including all dependencies

    ./mvnw clean install

### Checkstyle Config File

This project uses my [`common-parent`](https://github.com/patrickfav/mvn-common-parent) which centralized a lot of
the plugin versions aswell as providing the checkstyle config rules. Specifically they are maintained in [`checkstyle-config`](https://github.com/patrickfav/checkstyle-config). Locally the files will be copied after you `mvnw install` into your `target` folder and is called
`target/checkstyle-checker.xml`. So if you use a plugin for your IDE, use this file as your local configuration.

## Tech-Stack

* Java 7
* Maven (Wrapper)
* Errorprone, Checkstyle

## Related Libraries

* [BCyrpt Password Hash Function (Java)](https://github.com/patrickfav/bcrypt)
* [HKDF [RFC5869] Two-Step KDF (Java)](https://github.com/patrickfav/hkdf)

# License

Copyright 2018 Patrick Favre-Bulle

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG
# Releases

## v0.3.0

* use block length if null salt is provided instead of hash out length (makes no difference, hmac always pads to block length) #1
* add OSWAP dependency check plugin to Maven POM #2
* update various dependencies and move CI to github actions
* start using sonaqube

## v0.2.0

* add various unit tests
* update maven plugins

## v0.1.0

initial version


src/test/java/at/favre/lib/crypto/ConcatKdfTest.java
package at.favre.lib.crypto;

import at.favre.lib.bytes.Bytes;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class ConcatKdfTest {

    @Test
    public void compareAgainstRefValue() {
        byte[] ikm = Bytes.from("My secret").array();
        byte[] out = ConcatKdf.fromSha256().derive(ikm, null, new byte[16], 34);
        assertEquals(Bytes.wrap(out).encodeHex(), "5f225b4801843ed861b95f5b0a3afd78473498f0b5cb6d7769e67458e057da8c0311");
    }
}


src/test/java/at/favre/lib/crypto/HFunctionTest.java
package at.favre.lib.crypto;

import at.favre.lib.bytes.Bytes;
import org.junit.Test;

import javax.crypto.Mac;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import static org.junit.Assert.*;

public class HFunctionTest {

    @Test
    public void testMessageDigestHmacSha256() throws NoSuchAlgorithmException {
        HFunction digest = new HFunction.MacHFunction(Mac.getInstance("HmacSHA256"));
        assertTrue(digest.requireInit());
        assertEquals(32, digest.getHFuncOutputBytes());
        assertEquals(64, digest.getHFuncInputBlockLengthBytes());

        digest.init(new byte[32]);
        digest.update(new byte[0]);
        assertEquals("b613679a0814d9ec772f95d778c35fc5ff1697c493715653c6c712144292c5ad", Bytes.wrap(digest.calculate()).encodeHex());

        digest.reset();
    }

    @Test
    public void testMessageDigestHmacSha512() throws NoSuchAlgorithmException {
        HFunction digest = new HFunction.MacHFunction(Mac.getInstance("HmacSHA512"));
        assertTrue(digest.requireInit());
        assertEquals(64, digest.getHFuncOutputBytes());
        assertEquals(128, digest.getHFuncInputBlockLengthBytes());

        digest.init(new byte[64]);
        digest.update(new byte[0]);
        assertEquals("b936cee86c9f87aa5d3c6f2e84cb5a4239a5fe50480a6ec66b70ab5b1f4ac6730c6c515421b327ec1d69402e53dfb49ad7381eb067b338fd7b0cb22247225d47", Bytes.wrap(digest.calculate()).encodeHex());

        digest.reset();
    }

    @Test
    public void testMessageDigestSha256() throws NoSuchAlgorithmException {
        HFunction digest = new HFunction.MessageDigestHFunction(MessageDigest.getInstance("SHA-256"));
        assertFalse(digest.requireInit());
        assertEquals(32, digest.getHFuncOutputBytes());
        assertEquals(64, digest.getHFuncInputBlockLengthBytes());

        digest.update(new byte[0]);
        assertEquals("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", Bytes.wrap(digest.calculate()).encodeHex());

        digest.reset();
    }

    @Test
    public void testMessageDigestSha512() throws NoSuchAlgorithmException {
        HFunction digest = new HFunction.MessageDigestHFunction(MessageDigest.getInstance("SHA-512"));
        assertFalse(digest.requireInit());
        assertEquals(64, digest.getHFuncOutputBytes());
        assertEquals(128, digest.getHFuncInputBlockLengthBytes());

        digest.update(new byte[0]);
        assertEquals("cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e", Bytes.wrap(digest.calculate()).encodeHex());

        digest.reset();
    }

    @Test(expected = UnsupportedOperationException.class)
    public void testMessageDigestNotSupportInit() throws NoSuchAlgorithmException {
        new HFunction.MessageDigestHFunction(MessageDigest.getInstance("SHA-256")).init(new byte[32]);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testMacInvalidKey() throws NoSuchAlgorithmException {
        new HFunction.MacHFunction(Mac.getInstance("HmacSha256")).init(new byte[0]);
    }
}


src/test/java/at/favre/lib/crypto/SingleStepKdfTest.java
package at.favre.lib.crypto;

import at.favre.lib.bytes.Bytes;
import org.junit.Test;

import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.fail;

public class SingleStepKdfTest {

    @Test
    public void quickstart() {
        // a shared secret provided by your protocol
        byte[] sharedSecret = Bytes.random(16).array();
        // a salt; if you don't have access to a salt use SingleStepKdf.fromSha256() or similar
        byte[] salt = Bytes.random(16).array();
        // fixed info to bind the key to the context, see the NIST spec for more detail (previously otherInfo)
        byte[] fixedInfo = "macKey".getBytes();
        byte[] keyMaterial = SingleStepKdf.fromHmacSha256().derive(sharedSecret, 32, salt, fixedInfo);
        SecretKey secretKey = new SecretKeySpec(keyMaterial, "AES");
        assertNotNull(secretKey);
    }

    @Test
    public void quickstartMessageDigest() {
        // a shared secret provided by your protocol
        byte[] sharedSecret = Bytes.random(16).array();
        // fixed info to bind the key to the context
        byte[] fixedInfo = "macKey".getBytes();
        // salt is not directly supported with message digest, you may include it in fixedInfo
        byte[] keyMaterial = SingleStepKdf.fromSha256().derive(sharedSecret, 32, fixedInfo);
        SecretKey secretKey = new SecretKeySpec(keyMaterial, "AES");
        assertNotNull(secretKey);
    }

    @Test
    public void quickstartHmac() {
        byte[] salt = Bytes.random(16).array();
        // a shared secret provided by your protocol
        byte[] sharedSecret = Bytes.random(16).array();
        // fixed info to bind the key to the context
        byte[] fixedInfo = "macKey".getBytes();
        // salt is not directly supported with message digest, you may include it in fixedInfo
        byte[] keyMaterial = SingleStepKdf.fromHmacSha256().derive(sharedSecret, 32, salt, fixedInfo);
        SecretKey secretKey = new SecretKeySpec(keyMaterial, "AES");
        assertNotNull(secretKey);
    }

    @Test
    public void quickstartCustomImpl() {
        // create instance with sha1 as backing hash function
        SingleStepKdf sha1Kdf = SingleStepKdf.from(new HFunctionFactory.Default.DigestFactory("SHA-1"));
        // creat instance with HMAC-SHA1 as backing h function
        SingleStepKdf hmacSha1Kdf = SingleStepKdf.from(new HFunctionFactory.Default.DigestFactory("HmacSHA1"));


    }

    @Test
    public void testIllegalOutLength() {
        SingleStepKdf kdf = SingleStepKdf.fromSha256();
        try {
            kdf.derive(Bytes.random(16).array(), 0);
            fail();
        } catch (IllegalArgumentException ignored) {

        }
    }

    @Test
    public void testDoesNotSupportSalt() {
        SingleStepKdf kdf = SingleStepKdf.fromSha256();
        try {
            kdf.derive(Bytes.random(16).array(), 16, Bytes.random(16).array(), Bytes.random(16).array());
            fail();
        } catch (IllegalArgumentException ignored) {

        }
    }
}


src/main/java/at/favre/lib/crypto/singlstepkdf/HFunctionFactory.java
/*
 * Copyright 2018 Patrick Favre-Bulle
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package at.favre.lib.crypto.singlstepkdf;

import javax.crypto.Mac;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.Provider;

/**
 * Factory class for creating {@link MessageDigest}
 */

@SuppressWarnings({"unused", "WeakerAccess"})
public interface HFunctionFactory {

    /**
     * Creates a new instance of {@link MessageDigest}
     *
     * @return a new message digest instance
     */
    HFunction createInstance();

    /**
     * Get a user readable description of the used H-function (e.g. SHA-256 or HmacSha1 or similar)
     *
     * @return description
     */
    String getDescription();

    /**
     * Default implementation
     */
    final class Default {

        private Default() {
        }

        /**
         * Creates a factory creating SHA-256
         *
         * @return factory
         */
        public static HFunctionFactory sha256() {
            return new DigestFactory("SHA-256", null);
        }

        /**
         * Creates a factory creating SHA-512. Be aware that it is not guaranteed that this algorithm
         * is implemented in all JVMs.
         *
         * @return factory
         */
        public static HFunctionFactory sha512() {
            return new DigestFactory("SHA-512", null);
        }

        /**
         * Creates a factory creating HMAC with SHA-256
         *
         * @return factory
         */
        public static HFunctionFactory hmacSha256() {
            return new MacFactory("HmacSHA256", null);
        }

        /**
         * Creates a factory creating HMAC with SHA-512
         *
         * @return factory
         */
        public static HFunctionFactory hmacSha512() {
            return new MacFactory("HmacSHA512", null);
        }

        /**
         * Simple factory for message digests
         */
        public static final class DigestFactory implements HFunctionFactory {
            private final String algorithmName;
            private final Provider provider;

            /**
             * Creates a mac factory
             *
             * @param messageDigestAlgorithmName as used by {@link MessageDigest#getInstance(String)}
             */
            public DigestFactory(String messageDigestAlgorithmName) {
                this(messageDigestAlgorithmName, null);
            }

            /**
             * Creates a message digest factory
             *
             * @param messageDigestAlgorithmName as used by {@link MessageDigest#getInstance(String)}
             * @param provider                   what security provider, see {@link MessageDigest#getInstance(String, Provider)}; may be null to use default
             */
            public DigestFactory(String messageDigestAlgorithmName, Provider provider) {
                this.algorithmName = messageDigestAlgorithmName;
                this.provider = provider;
            }

            @Override
            public HFunction createInstance() {
                try {
                    MessageDigest messageDigest;

                    if (provider == null) {
                        messageDigest = MessageDigest.getInstance(algorithmName);
                    } else {
                        messageDigest = MessageDigest.getInstance(algorithmName, provider);
                    }

                    return new HFunction.MessageDigestHFunction(messageDigest);
                } catch (NoSuchAlgorithmException e) {
                    throw new IllegalStateException("defined message digest algorithm was not found", e);
                }
            }

            @Override
            public String getDescription() {
                return "MessageDigest[" + algorithmName + "]";
            }
        }

        /**
         * Simple factory for MAC
         */
        public static final class MacFactory implements HFunctionFactory {
            private final String algorithmName;
            private final Provider provider;

            /**
             * Creates a mac factory
             *
             * @param macAlgorithmName as used by {@link Mac#getInstance(String)}
             */
            public MacFactory(String macAlgorithmName) {
                this(macAlgorithmName, null);
            }

            /**
             * Creates a message digest factory
             *
             * @param macAlgorithmName as used by {@link Mac#getInstance(String)}
             * @param provider         what security provider, see {@link Mac#getInstance(String, Provider)}; may be null to use default
             */
            public MacFactory(String macAlgorithmName, Provider provider) {
                this.algorithmName = macAlgorithmName;
                this.provider = provider;
            }

            @Override
            public HFunction createInstance() {
                try {
                    Mac mac;

                    if (provider == null) {
                        mac = Mac.getInstance(algorithmName);
                    } else {
                        mac = Mac.getInstance(algorithmName, provider);
                    }

                    return new HFunction.MacHFunction(mac);
                } catch (NoSuchAlgorithmException e) {
                    throw new IllegalStateException("defined mac algorithm was not found", e);
                }
            }

            @Override
            public String getDescription() {
                return "MAC[" + algorithmName + "]";
            }
        }
    }
}


src/main/java/at/favre/lib/crypto/singlstepkdf/HFunction.java
/*
 * Copyright 2018 Patrick Favre-Bulle
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package at.favre.lib.crypto.singlstepkdf;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.util.Objects;

/**
 * The auxiliary function used to produce blocks of keying
 * material during the execution of the one-step key-derivation
 * method. Basically an abstraction of either {@link MessageDigest} or {@link Mac}
 * <p>
 * Options for the Auxiliary Function H:
 * <ul>
 * <li>hash(x), where hash is an approved hash function</li>
 * <li>HMAC-hash(salt, x), where HMAC-hash is an implementation of the HMAC algorithm</li>
 * <li>KMAC#(salt, x, H_outputBits, S), where KMAC# is a particular implementation of either KMAC128 or KMAC256 </li>
 * </ul>
 * <p>
 * An example on how to use a H-function
 * <pre>
 *     final HFunction func = factory.createInstance();
 *     final int inputBlockLength = func.getHFuncInputBlockLengthBytes();
 *     byte[] out = new byte[hashLength];
 *     func.update(keyData);
 *     out = func.calculate()
 * </pre>
 * <p>
 * see <a href="https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr1.pdf">NIST.SP.800-56Cr1</a>
 */
@SuppressWarnings("WeakerAccess")
public interface HFunction {

    /**
     * Optional keying of the H function. This is not supported
     * with hash type functions and may throw an {@link UnsupportedOperationException}. Call this before {@link #update(byte[])}
     *
     * @param key to be used; if null is passed, default_salt is used, which is in the case of
     *            HMAC an empty byte array the length of the underyling hash function (e.g. 32
     *            bytes for SHA-256)
     */
    void init(byte[] key);

    /**
     * If the function requires a key to be provided in the {@link #init(byte[])}
     *
     * @return true if user must initialize with init()
     */
    boolean requireInit();

    /**
     * Formally: a positive integer that indicates the length of the input block size of
     * the auxiliary function, H.
     * <p>
     * The spec describes the length in bits, but for practical reasons it will return bytes.
     * <p>
     * This is required to get the salt length of no salt is provided.
     * See the definition in NIST SP 800-56C REV. 1, Table 2
     *
     * @return input block size length in byte (sha-256 is e.g. 64, sha-512: 128)
     */
    int getHFuncInputBlockLengthBytes();

    /**
     * Formally: a positive integer that indicates the length of the output of
     * the auxiliary function, H, that is used to derive blocks of secret keying material.
     * <p>
     * The spec describes the length in bits, but for practical reasons it will return bytes.
     * <p>
     * Practically the length of output hash/hmac/kmac function. E.g. SHA-256 or HMAC_SHA-256
     * will output 32 byte long block.
     *
     * @return output length in byte
     */
    int getHFuncOutputBytes();

    /**
     * Update the function with a new input that will internally add to the
     * already available input
     *
     * @param array to process
     */
    void update(byte[] array);

    /**
     * After adding the input through {@link #update(byte[])} calculate the output of the
     * H-function
     *
     * @return output (usually a hash)
     */
    byte[] calculate();

    /**
     * Resets the function for further use.
     */
    void reset();

    /**
     * Hash implementation of the H function.
     */
    final class MessageDigestHFunction implements HFunction {
        private final MessageDigest digest;

        public MessageDigestHFunction(MessageDigest digest) {
            this.digest = Objects.requireNonNull(digest, "digest");
        }

        @Override
        public void init(byte[] key) {
            throw new UnsupportedOperationException("message digest does not support keying");
        }

        @Override
        public boolean requireInit() {
            return false;
        }

        @Override
        public int getHFuncInputBlockLengthBytes() {
            return Util.getBlockLengthByte(digest.getAlgorithm());
        }

        @Override
        public int getHFuncOutputBytes() {
            return digest.getDigestLength();
        }

        @Override
        public void update(byte[] array) {
            digest.update(array);
        }

        @Override
        public byte[] calculate() {
            return digest.digest();
        }

        @Override
        public void reset() {
            digest.reset();
        }
    }

    /**
     * HMAC implementation of the H function
     */
    final class MacHFunction implements HFunction {
        private final Mac mac;

        public MacHFunction(Mac mac) {
            this.mac = Objects.requireNonNull(mac, "mac");
        }

        @Override
        public void init(byte[] key) {
            /*
             * If a salt value required by H is omitted from OtherInput (or if a required salt value
             * included in OtherInput is the null string), then the value of default_salt shall be
             * used as the value of salt when H is executed.
             *
             * default_salt: a non-null (secret or non-secret) byte string that is needed only if
             * either Option 2 (HMAC-hash) or Option 3 (KMAC#) is chosen for the implementation
             * of the auxiliary function H. This byte string is used as the value of salt if a
             * (non-null) value is not included in OtherInput. If H(x) = HMAC-hash(salt, x), then,
             * in the absence of an agreed-upon alternative, the default_salt shall be an all-zero
             * byte string whose bit length equals that specified as the bit length of an input
             * block for the hash function, hash.
             */
            if (key == null) {
                key = new byte[getHFuncInputBlockLengthBytes()];
            }

            try {
                mac.init(new SecretKeySpec(key, mac.getAlgorithm()));
            } catch (InvalidKeyException e) {
                throw new IllegalArgumentException("invalid key", e);
            }
        }

        @Override
        public boolean requireInit() {
            return true;
        }

        @Override
        public int getHFuncInputBlockLengthBytes() {
            return Util.getBlockLengthByte(mac.getAlgorithm());
        }

        @Override
        public int getHFuncOutputBytes() {
            return mac.getMacLength();
        }

        @Override
        public void update(byte[] array) {
            mac.update(array);
        }

        @Override
        public byte[] calculate() {
            return mac.doFinal();
        }

        @Override
        public void reset() {
            mac.reset();
        }
    }

    final class Util {
        private Util() {
        }

        static int getBlockLengthByte(String algorithm) {
            String name = Objects.requireNonNull(algorithm).toLowerCase().trim().replace("-", "");

            if (name.startsWith("sha1") || name.startsWith("sha224") || name.startsWith("sha256")
                    || name.startsWith("hmacsha1") || name.startsWith("hmacsha256")) {
                return 64;
            } else if (name.startsWith("sha512") || name.startsWith("sha384") || name.startsWith("hmacsha512")) {
                return 128;
            } else if (name.startsWith("sha3224")) {
                return 144;
            } else if (name.startsWith("sha3256")) {
                return 136;
            } else if (name.startsWith("sha3384")) {
                return 104;
            } else if (name.startsWith("sha3512")) {
                return 72;
            } else {
                throw new IllegalStateException("unknown hash algorithm; cannot choose input block length: see NIST SP 800-56C REV. 1 Table 2.");
            }
        }
    }
}
